instance/models/instance.py

```diff
@@ -0,0 +1,476 @@
+            self.branch_name,
+            ref_type=self.ref_type)
+        if commit:
+            self.save()
+
+    def set_fork_name(self, fork_name, commit=True):
+        """
+        Set the organization and repository based on the GitHub fork name
+
+        By default, save the instance object - pass `commit=False` to not save it
+        """
+        self.log('info', 'Setting fork name for instance {}: {}'.format(self, fork_name))
+        fork_org, fork_repo = github.fork_name2tuple(fork_name)
+        if self.github_organization_name == fork_org \
+                and self.github_repository_name == fork_repo:
+            return
+
+        self.github_organization_name = fork_org
+        self.github_repository_name = fork_repo
+        if commit:
+            self.save()
+
+
+# Ansible #####################################################################
+
+class AnsibleInstanceMixin(models.Model):
+    """
+    An instance that relies on Ansible to deploy its services
+    """
+    ansible_source_repo_url = models.URLField(max_length=256,
+                                              default='https://github.com/edx/configuration.git')
+
+    configuration_version = models.CharField(max_length=50, default='master')
+    ansible_playbook_name = models.CharField(max_length=50, default='edx_sandbox')
+    ansible_extra_settings = models.TextField(blank=True)
+
+    # List of attributes to include in the settings output
+    ANSIBLE_SETTINGS = ['ansible_extra_settings']
+
+    class Meta:
+        abstract = True
+
+    @property
+    def ansible_playbook_filename(self):
+        """
+        File name of the ansible playbook
+        """
+        return '{}.yml'.format(self.ansible_playbook_name)
+
+    @property
+    def inventory_str(self):
+        """
+        The ansible inventory (list of servers) as a string
+        """
+        inventory = ['[app]']
+        server_model = self.server_set.model
+        for server in self.server_set.filter(status=server_model.BOOTED).order_by('created'):
+            inventory.append(server.public_ip)
+        inventory_str = '\n'.join(inventory)
+        self.log('debug', 'Inventory for instance {}:\n{}'.format(self, inventory_str))
+        return inventory_str
+
+    @property
+    def vars_str(self):
+        """
+        The ansible vars (private configuration) as a string
+        """
+        template = loader.get_template('instance/ansible/vars.yml')
+        vars_str = template.render({'instance': self})
+        for attr_name in self.ANSIBLE_SETTINGS:
+            additional_vars = getattr(self, attr_name)
+            vars_str = ansible.yaml_merge(vars_str, additional_vars)
+        self.log('debug', 'Vars.yml for instance {}:\n{}'.format(self, vars_str))
+        return vars_str
+
+    def run_playbook(self, playbook_name=None):
+        """
+        Run a playbook against the instance active servers
+        """
+        with clone_configuration_repo() as configuration_repo_path:
```

Comment:
@bradenmacdonald Done - as part of refactoring needed for the integration tests (which needed to fetch a different repo & branch), I reworked this section and made it fully generic; it's now `with open_repository(repo_url, ref=branch) as repo: ...`